dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,159 @@
+manage shared memory used to pass results of the callback calls.
+"""
+
+    def __init__(self, callback, mem_chunks):
+        self.callback = callback
+        self.mem_chunks = mem_chunks
+        self.batch_i = 0
+
+    def next_batch_i(self):
+        res = self.batch_i
+        self.batch_i = (self.batch_i + 1) % len(self.mem_chunks)
+        return res
+
+    def next_mem_chunk(self):
+        batch_i = self.next_batch_i()
+        return self.mem_chunks[batch_i]
+
+    def close(self):
+        for chunk in self.mem_chunks:
+            chunk.close()
+
+
+def dispatcher(batch_dispatcher, ready_cv, ready_queue):
+    """Receives batches produced in the main thread and dispatches them to the parent process.
+It's run in separate thread for two reasons: one is that both callback and dispatcher may
+wait on IO operations a lot and in that case Python threads provide some performance gain.
+Secondly it prevents main thread from blocking on sending the data (so that we don't
+deadlock when parent process schedules too much tasks, waits trying to send another task
+and worker process doesn't read from pipe because it waits on writing to results pipe).
+"""
+    while True:
+        with ready_cv:
+            while len(ready_queue) == 0:
+                ready_cv.wait()
+            message = ready_queue.pop(0)
+        if message is None:
+            break
+        batch_dispatcher.send(*message)
+
+
+def worker(worker_id, callbacks, prefetch_queue_depths, initial_chunk_size, task_pipe, res_pipe, sock):
+    """Entry point of worker process.
+
+Parameters
+----------
+`callbacks` : callable list
+    List of callables that worker can call to perform a (part of parallelized) task.
+`prefetch_queue_depths` : int
+    Number of shared memory chunks that should be allocated per callaback, used in cycle buffer manner
+    to pass callback results to parent process.
+`initial_chunk_size` : int
+    Initial size of shared memory chunk.
+`task_pipe`: Pipe
+    Pipe used to read list of tasks that given callback should be run on to produce (part of a) result batch.
+`res_pipe`: Pipe
+    Pipe used to notify the parent process about another batch ready to read in the given memory chunk.
+`sock` : socket
+    Python wrapper around Unix socket used to pass file descriptors identifying shared memory chunk to parent process.
+"""
+    contexts = None
+    ready_cv = threading.Condition()
+    ready_queue = []
+    batch_dispatcher = SharedBatchesDispatcher(worker_id, sock, res_pipe)
+    # run the thread as a daemon so that even when results queue blocks worker process can exit anyway
+    # and can be joined in the parent process
+    dispatcher_thread = threading.Thread(target=dispatcher, args=(batch_dispatcher, ready_cv, ready_queue), daemon=True)
+    dispatcher_thread.start()
+    try:
+        contexts = [
+            CallbackContext(callback, [
+                SharedMemChunk("chunk_{}_{}_{}".format(worker_id, i, j), initial_chunk_size)
```

Comment:
How this `"chunk_{}_{}_{}".format(worker_id, i, j)` is used. String looks only as a good thing for debugging.

Reply:
It is used as an id of shared memory chunk. When parent process receives batch it checks if it ever received batch with given memory chunk id, if not it knows to expect file descriptor for new shared memory chunk.

Reply:
ok